### ISSUE TITLE
Catalog readonly mode allows delete of entity

### ISSUE THREAD

**Issue by pschilli** [backstage/backstage#12146]
## Expected Behavior

When the catalog is in readonly mode, a `DELETE` on ` /api/catalog/entities/by-uid/{uuid}` should not be allowed.

## Actual Behavior

`DELETE` on ` /api/catalog/entities/by-uid/{uuid}` is allowed.

` disallowReadonlyMode` is not currently applied to the route https://github.com/backstage/backstage/blob/d14c1059f7e734ae9feaaad90f04420fa8643ae2/plugins/catalog-backend/src/service/createRouter.ts#L141-L147

## Steps to Reproduce

1. Configure readonly mode for the catalog
```
catalog:
  readonly: true
```
2. Execute a `DELETE` against an entity URI with its UUID `/api/catalog/entities/by-uid/{uuid}`
3. The entity is removed, and will be missing until discovered again.

## Context

I am currently blocking the route when in readonly mode. 
I want to understand if allowing the delete is intentional behavior for readonly mode.

## Your Environment

- Browser Information:  n/a, backend API

- Output of `yarn backstage-cli info`: 

```text
OS:   Darwin 20.6.0 - darwin/x64
node: v16.14.0
yarn: 1.22.17
cli:  0.17.1 (installed)

Dependencies:
  @backstage/app-defaults                          1.0.2
  @backstage/backend-common                        0.13.3
  @backstage/backend-tasks                         0.3.1
  @backstage/catalog-client                        1.0.2
  @backstage/catalog-model                         1.0.2
  @backstage/cli-common                            0.1.9
  @backstage/cli                                   0.17.1
  @backstage/config-loader                         1.1.1
  @backstage/config                                1.0.1
  @backstage/core-app-api                          1.0.2
  @backstage/core-components                       0.9.4
  @backstage/core-plugin-api                       1.0.2
  @backstage/errors                                1.0.0
  @backstage/integration-react                     1.1.0
  @backstage/integration                           1.2.0
  @backstage/plugin-api-docs                       0.8.5
  @backstage/plugin-app-backend                    0.3.32
  @backstage/plugin-auth-backend                   0.14.0
  @backstage/plugin-auth-node                      0.2.1
  @backstage/plugin-catalog-backend-module-github  0.1.3
  @backstage/plugin-catalog-backend                1.1.2
  @backstage/plugin-catalog-common                 1.0.2
  @backstage/plugin-catalog-graph                  0.2.17
  @backstage/plugin-catalog-import                 0.8.8
  @backstage/plugin-catalog-react                  1.1.0
  @backstage/plugin-catalog                        1.2.0
  @backstage/plugin-github-actions                 0.5.5
  @backstage/plugin-org                            0.5.5
  @backstage/plugin-permission-common              0.6.1
  @backstage/plugin-permission-node                0.6.1
  @backstage/plugin-permission-react               0.4.1
  @backstage/plugin-proxy-backend                  0.2.26
  @backstage/plugin-scaffolder-backend             1.2.0
  @backstage/plugin-scaffolder-common              1.1.0
  @backstage/plugin-scaffolder                     1.2.0
  @backstage/plugin-search-backend-module-pg       0.3.3
  @backstage/plugin-search-backend-node            0.6.1
  @backstage/plugin-search-backend                 0.5.2
  @backstage/plugin-search-common                  0.3.4
  @backstage/plugin-search-react                   0.2.0
  @backstage/plugin-search                         0.8.1
  @backstage/plugin-tech-radar                     0.5.12
  @backstage/plugin-techdocs-backend               1.1.1
  @backstage/plugin-techdocs-module-addons-contrib 1.0.0
  @backstage/plugin-techdocs-node                  1.1.1
  @backstage/plugin-techdocs-react                 1.0.0
  @backstage/plugin-techdocs                       1.1.1
  @backstage/plugin-user-settings                  0.4.4
  @backstage/release-manifests                     0.0.3
  @backstage/search-common                         0.3.4
  @backstage/test-utils                            1.1.0
  @backstage/theme                                 0.2.15
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1

```

**Comment by Rugvip** [comment]
The read only mode is specifically for locations and whether users are allowed to register and unregister locations.

The delete endpoint is for a slightly different use-case, removing [orphaned](https://backstage.io/docs/features/software-catalog/life-of-an-entity#orphaning) entities. Entities can be orphaned regardless of whether the catalog is in read only mode or not. To be honest I think - to your point - that it is a bit confusing to allow deletion of non-orphaned entities. It's something we've debated quite a bit, with the argument for it being that you might end up with entities being stuck in the catalog for some reason, and then you want to have a way to delete them.

To perhaps get some action going I figure we could perhaps go for some voting:

| Reaction | Vote |
| :-: | - |
| 🚀 | I think deletions should only be allowed for orphaned entities, because it is confusing that they can be removed, even if they reappear later. |
| 🎉 | I think it should be possible to delete all entities since it is a good safeguard and utility to resolve possible issues with the catalog processing loop. |
| 👀 | Just wanna let you know that I've seen this but either outcome is alright by me. |

Hoping I represented all sides fairly here 😅, react with the emoji matching your vote. Or comment if I missed something.

**Comment by ozialien** [comment]
Isn't this an RBAC issue also.  Entities with enough privilege should be able to delete entities via some RBAC scheme.  May be complicating the immediate issue though.

**Comment by pschilli** [comment]
Thanks for the context, the orphan delete was the piece I was missing.  I think it would be helpful to be able to disable delete of non-orphaned entities while still allowing delete of orphaned ones.  Is there a third option where it's configurable?  😅

**Comment by ericis** [comment]
Suggest an action-based ReST URI for "orphan cleanup" on the "entities" resource collection. Currently, the URI "/api/catalog/entities/by-uid/{uuid}" includes:
- the "catalog" resource (technically a shortcut to a resource "by name" in the "plugins" resource collection)
- the "entities" resource collection
- A path implementation of what could be a query filter for "by uid" along with the parameter value

A proposed solution could be to add something like:

- "/api/catalog/entities/clean", or
- "/api/catalog/entities/clean?uid={uuid}", or
- "/api/catalog/entities/orphans/", or
- "/api/catalog/entities/orphans?uid={uuid}", or
- "/api/catalog/entities/orphans/by-uid/{uuid}"

Although, I might also suggest "/api/plugins/catalog" for the future as I believe the API is for a "Backstage API" holistically and not for just a "Backstage Plugins API", but I could be mistaken if every service is in support of a backend plugin.

**Comment by freben** [comment]
Not configurable at the moment, no. Maybe this is the point where the permissions framework should be leveraged instead, to disable deletes perhaps with the exception of for admins. Disabling it entirely would probably be a mistake, because the functionality it provides is usually relevant to have at some times. The readonly flag, in my mind, is a remnant from the past - way before the permissions framework existed - and it may be the case that it should be retired entirely. Not sure that everybody agrees though :)

**Comment by freben** [comment]
@ericis I don't think there's any particular value in duplicating all of the `by-uid` endpoints. And the only reason they exist at all, in addition to the ability to do `/entities?filter=`, is only for convenience of easily getting either a single item or a 404. They aren't strictly necessary at all, of course.

Searching for orphans is already possible (just filtering for that annotation being present).

I wouldn't call `catalog` a resource collection, that's not really the mindset or underpinnings of that path structure. This all boils down to

- the ability to co-host the frontend and the backend
- the ability to run multiple backends on a single host
- how our discovery api is set up to deal with the above

The `/api` prefix is there (note - in the DEFAULT setup - adopters can do anything they want) only to be able to clearly distinguish backend-destined traffic from frontend-destined traffic. In our own deployment we have a small ingress proxy that does incoming traffic splitting based on that. Essentially, your `/api/plugins/catalog` is exactly what we do, just that it's only `/api/catalog` because there's no other "layer" to be had in between the two.

The `/catalog` suffix is there (note - again, in the DEFAULT setup, this is in no way enforced) to make the freshly scaffolded backstage repo able to co-host many backend plugins on a single deployment, and the standard then is to use the plugin ID as a path parameter. Could just as well have been a port mapping instead in the default setup, and a custom setup may resolve to entirely different domains even for different plugins.

So the only resource collection at play is really `/entities`. :)

**Comment by ericis** [comment]
Yes, I was indeed suggesting "catalog" is implicitly a "plugin" resource in the "plugins" resource collection and that this question about "cleanup" is an action on the "entities" resource collection when performed in bulk. And,  there still should be a "DELETE" verb for a single "entity" in "entities" and that, usually, the "purist" ReST approach to filtering is through the querystring and not in the path of the URL (e.g. "by-uuid").

@freben you've provided great context to the intentionality behind the current implementation.

**Comment by pschilli** [comment]
The docs are a bit confusing to a newcomer.  Would it make sense to add some additional detail to https://backstage.io/docs/features/software-catalog/configuration#readonly-mode?  The `disables the mutating backend catalog APIs` is the part that doesn't quite agree with the current behavior. I would consider a delete to be mutating, even if it's rediscovered.

The permissions framework seems like the right way forward for restricting delete, especially if `readonly` is or will soon be considered deprecated. The `catalog.entity.delete` permission is defined for this use.  If `readonly` will continue forward, it might make sense to look at an option to disable delete of non-orphan entities.

**Comment by freben** [comment]
Yeah a doc update there would totally make sense, want to take a stab at a better formulation in a PR?

And agreed, I think there are already some permissions related to adding/removing entities to there may be some redundancy here now after that framework was introduced

**Comment by pschilli** [comment]
I made an attempt at a doc update with #12301

**Comment by freben** [comment]
Closing since the doc update is merged!
